server/src/database/remoteStorage.ts

```diff
@@ -0,0 +1,128 @@
+import {
+  S3Client,
+  ListBucketsCommand,
+  ListObjectsV2Command,
+  GetObjectCommand,
+  PutObjectCommand,
+  type ListBucketsCommandOutput,
+  type ListObjectsV2CommandOutput,
+  type GetObjectCommandOutput,
+  type PutObjectCommandOutput,
+  type S3ServiceException,
+} from '@aws-sdk/client-s3'
+
+export class StoreFile<T> {
+  private readonly directory!: string
+  private readonly storage!: IStorage
+
+  constructor (directory: string, storage: IStorage) {
+    this.directory = directory
+    this.storage = storage
+  }
+
+  getFile: (fileName: string) => Promise<T | null> = async (fileName: string) => await this.storage.getFile(this.directory, fileName)
+  putFile: (fileName: string, object: T) => Promise<void> = async (fileName: string, object: T) => { await this.storage.putFile(this.directory, fileName, object) }
+}
+
+interface IStorage {
+  listDirectories: () => Promise<string[]>
+  listFiles: (directory: string) => Promise<string[]>
+  getFile: <T>(directory: string, file: string) => Promise<T | null>
+  putFile: <T>(directory: string, file: string, object: T) => Promise<void>
+}
+
+export class RemoteStorage implements IStorage {
+  private readonly storage: S3Storage
+
+  constructor () {
+    this.storage = new S3Storage()
+  }
+
+  listDirectories = async (): Promise<string[]> => {
+    const response = await this.storage.listBuckets()
+    if (response.$metadata.httpStatusCode === 200 && response.Buckets) {
+      return response.Buckets.map(({ Name }) => Name as string)
+    }
+
+    throw new Error('Remote storage error on listBuckets')
+  }
+
+  listFiles = async (directory: string): Promise<string[]> => {
+    const response = await this.storage.listObjects(directory)
+    if (response.$metadata.httpStatusCode === 200 && response.Contents) {
+      return response.Contents.map(({ Key }) => Key as string)
+    }
+
+    throw new Error('Remote storage error on listObjects')
+  }
+
+  getFile = async <T>(directory: string, file: string): Promise<T | null> => {
+    try {
+      const response = await this.storage.getObject(directory, file)
+      if (response.$metadata.httpStatusCode === 200 && response.Body) {
+        return JSON.parse(await response.Body.transformToString()) as T
+      }
+    } catch (e) {
+      const err = e as S3ServiceException
+      if (err.name === 'NoSuchKey') return null
```

Comment:
Hacky way to check if file exists.
S3 also exposes a "get metadata" endpoint which is the closest to "has file" we have, but it also throws 'NoSuchKey' when file not found.  So I'm not bothering to check, just grab if there